### PR TITLE
Add a flat l2 topology NAD option

### PR DESCRIPTION
**Pull request by Anatw**
##### Short description:
Add a flat L2 network on OVN Kubernetes.

##### What this PR does / why we need it:
Support a new feature.

### CODE REVIEW

File: ocp_resources/network_attachment_definition.py [hunk 1]
```diff
@@ -197,3 +197,48 @@ def to_dict(self):
     @property
     def resource_name(self):
         return f"ovs-cni.network.kubevirt.io/{self.bridge_name}"
+
+
+class FlatL2NetworkAttachmentDefinition(NetworkAttachmentDefinition):
```
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:202]: Ack.

File: ocp_resources/network_attachment_definition.py [hunk 2]
```diff
@@ -197,3 +197,48 @@ def to_dict(self):
...
+    def __init__(
+        self,
+        name,
```
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:205]: Ack.

File: ocp_resources/network_attachment_definition.py [hunk 3]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
     @property
     def resource_name(self):
         return f"ovs-cni.network.kubevirt.io/{self.bridge_name}"
+
+
+class FlatL2NetworkAttachmentDefinition(NetworkAttachmentDefinition):
+    def __init__(
+        self,
+        network_name=None,
+        teardown=True,
```
Comment (myakove) [inline ocp_resources/network_attachment_definition.py:206]: `teardown` should not be here

File: ocp_resources/network_attachment_definition.py [hunk 4]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
     @property
     def resource_name(self):
         return f"ovs-cni.network.kubevirt.io/{self.bridge_name}"
+
+
+class FlatL2NetworkAttachmentDefinition(NetworkAttachmentDefinition):
```
Comment (myakove) [inline ocp_resources/network_attachment_definition.py:202]: Is this resource is standalone resource? or extended NetworkAttachmentDefinition ?

File: ocp_resources/network_attachment_definition.py [hunk 5]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+        self.topology = "layer2"
+        self.netAttachDefName = f"{self.namespace}/{self.name}"
+        self.network_name = (
+            f"{network_name}-network" if network_name is None else network_name
+        )
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            spec_config = self.res["spec"]["config"]
```
Comment (myakove) [inline ocp_resources/network_attachment_definition.py:234]: If some args are mandatory you need to assert them if not exists

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/network_attachment_definition.py [hunk 6]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+    def __init__(
+        self,
+        network_name=None,
+        teardown=True,
+        *args,
+        **kwargs,
+    ):
+        """
+        Create and manage a flat-l2 NetworkAttachmentDefinition.
+        API reference:
```
Comment (rnetser) [inline ocp_resources/network_attachment_definition.py:212]: - please update docstring, it is not `flat-l2` 
- please use proper docstring formatting

File: ocp_resources/network_attachment_definition.py [hunk 7]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+            teardown=teardown,
+            *args,
+            **kwargs,
+        )
+        self.topology = "layer2"
+        self.netAttachDefName = f"{self.namespace}/{self.name}"
+        self.network_name = (
+            f"{network_name}-network" if network_name is None else network_name
+        )
+
```
Comment (rnetser) [inline ocp_resources/network_attachment_definition.py:230]:
```
In [1]: network_name=None

In [2]: (
   ...:             f"{network_name}-network" if network_name is None else network_name
   ...:         )
Out[2]: 'None-network'
```
this name does not make sense

File: ocp_resources/network_attachment_definition.py [hunk 8]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+            network_name (str, default: None): The name of the network. This attribute is not namespaced.
+            teardown (bool, default: True): Indicates if this resource would need to be deleted.
+        """
+        super().__init__(
+            cni_type="ovn-k8s-cni-overlay",
+            teardown=teardown,
+            *args,
+            **kwargs,
+        )
+        self.topology = "layer2"
```
Comment (rnetser) [inline ocp_resources/network_attachment_definition.py:225]: used once, no need to save

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/network_attachment_definition.py (hunk 4 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:202]: It is a NetworkAttachmentDefinition.

File: ocp_resources/network_attachment_definition.py (hunk 5 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:234]: There is an assertion for that in the base class NamespacedResource.

File: ocp_resources/network_attachment_definition.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:202]: Ack

File: ocp_resources/network_attachment_definition.py [hunk 9]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+        *args,
+        **kwargs,
+    ):
+        """
+        Create and manage a flat-l2 NetworkAttachmentDefinition.
+        API reference:
+        https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/configuring-additional-network.html#configuration-ovnk-network-plugin-json-object_configuring-additional-network
+
+        Args:
+            network_name (str, default: None): The name of the network. This attribute is not namespaced.
```
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:216]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 7 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:230]: Currect, thanks! I meant to use "name" and not "network_name" in the beginning. It doesn't make sense here now that we don't pass the name.
I'll refactor and move it to the to_dict function.

File: ocp_resources/network_attachment_definition.py (hunk 6 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:212]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 8 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:225]: Ack

File: ocp_resources/network_attachment_definition.py [hunk 10]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+            teardown (bool, default: True): Indicates if this resource would need to be deleted.
+        """
+        super().__init__(
+            cni_type="ovn-k8s-cni-overlay",
+            teardown=teardown,
+            *args,
+            **kwargs,
+        )
+        self.topology = "layer2"
+        self.netAttachDefName = f"{self.namespace}/{self.name}"
```
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:226]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 5 above)
Comment (rnetser) [reply ocp_resources/network_attachment_definition.py:234]: `network_name` should be mandatory and not set for the user

File: ocp_resources/network_attachment_definition.py [hunk 11]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+
+        Args:
+            network_name (str, default: None): The name of the network.
+        """
+        super().__init__(
+            cni_type="ovn-k8s-cni-overlay",
+            *args,
+            **kwargs,
+        )
+        self.net_attach_def_name = f"{self.namespace}/{self.name}"
```
Comment (rnetser) [inline ocp_resources/network_attachment_definition.py:223]: used once

File: ocp_resources/network_attachment_definition.py [hunk 12]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
     @property
     def resource_name(self):
         return f"ovs-cni.network.kubevirt.io/{self.bridge_name}"
+
+
+class OvnOverlayNetworkAttachmentDefinition(NetworkAttachmentDefinition):
```
Comment (rnetser) [inline ocp_resources/network_attachment_definition.py:202]:
```suggestion
class OVNOverlayNetworkAttachmentDefinition(NetworkAttachmentDefinition):
```

File: ocp_resources/network_attachment_definition.py [hunk 13]
```diff
@@ -197,3 +197,42 @@ def to_dict(self):
...
+        **kwargs,
+    ):
+        """
+        Create and manage an OVN k8s overlay NetworkAttachmentDefinition (a switched, layer 2, topology network).
+
+        API reference:
+        https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/configuring-additional-network.html#configuration-ovnk-network-plugin-json-object_configuring-additional-network
+
+        Args:
+            network_name (str, default: None): The name of the network.
```
Comment (rnetser) [inline ocp_resources/network_attachment_definition.py:216]:
```suggestion
            network_name (str): The name of the network.
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/network_attachment_definition.py (hunk 11 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:223]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 13 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:216]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 12 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:202]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 5 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:234]: Ack

File: ocp_resources/network_attachment_definition.py [hunk 14]
```diff
@@ -197,3 +197,36 @@ def to_dict(self):
     @property
     def resource_name(self):
         return f"ovs-cni.network.kubevirt.io/{self.bridge_name}"
+
+
+class OVNOverlayNetworkAttachmentDefinition(NetworkAttachmentDefinition):
+    def __init__(
+        self,
+        network_name,
```
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:205]: Ack.

File: ocp_resources/network_attachment_definition.py [hunk 15]
```diff
@@ -197,3 +197,39 @@ def to_dict(self):
     @property
     def resource_name(self):
         return f"ovs-cni.network.kubevirt.io/{self.bridge_name}"
+
+
+class OVNOverlayNetworkAttachmentDefinition(NetworkAttachmentDefinition):
+    def __init__(
+        self,
+        network_name=None,
+        *args,
```
Comment (myakove) [inline ocp_resources/network_attachment_definition.py:206]: we limit it only to `kwargs`, remove `args`

File: ocp_resources/network_attachment_definition.py (hunk 6 above)
Comment (myakove) [reply ocp_resources/network_attachment_definition.py:212]: Also update the PR title

File: ocp_resources/network_attachment_definition.py [hunk 16]
```diff
@@ -197,3 +197,39 @@ def to_dict(self):
...
+        super().__init__(
+            cni_type="ovn-k8s-cni-overlay",
+            *args,
+            **kwargs,
+        )
+        self.network_name = network_name
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.network_name:
```
Comment (myakove) [inline ocp_resources/network_attachment_definition.py:228]: If user send `yaml_file` and not  `network_name` I will fail while it shouldn't

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/network_attachment_definition.py (hunk 15 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:206]: Ack

File: ocp_resources/network_attachment_definition.py (hunk 16 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py:228]: Ack.

File: ocp_resources/network_attachment_definition.py [hunk 17]
```diff
@@ -197,3 +197,37 @@ def to_dict(self):
...
+        **kwargs,
+    ):
+        """
+        Create and manage an OVN k8s overlay NetworkAttachmentDefinition (a switched, layer 2, topology network).
+
+        API reference:
+        https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/configuring-additional-network.html#configuration-ovnk-network-plugin-json-object_configuring-additional-network
+
+        Args:
+            network_name (str, default: None): The name of the network. This field is mandatory when not using yaml
```
Comment (myakove) [inline ocp_resources/network_attachment_definition.py:215]: `default: None` please remove (you set default in docstring when you have default, not None)

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/network_attachment_definition.py (hunk 17 above)
Comment (Anatw) [reply ocp_resources/network_attachment_definition.py]: Ack

**Review by myakove — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Pull request merged**